lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -842,45 +870,80 @@ namespace Js
     }
 
 #ifdef INTL_ICU
-    typedef const char * (__cdecl *GetAvailableLocaleFunc)(int);
-    typedef int (__cdecl *CountAvailableLocaleFunc)(void);
-    static bool findLocale(JavascriptString *langtag, CountAvailableLocaleFunc countAvailable, GetAvailableLocaleFunc getAvailable)
+    template <const char *(__cdecl *GetAvailableLocalesFunc)(int), int(__cdecl *CountAvailableLocalesFunc)(void)>
+    static bool BinarySearchForLocale(const char *localeID)
     {
-        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
-        BCP47_TO_ICU(langtag->GetSz(), langtag->GetLength(), localeID, ULOC_FULLNAME_CAPACITY);
-
-        // ICU's "available locales" do not include (all? most?) aliases.
-        // For example, searching for "zh_TW" will return false, even though
-        // zh_TW is equivalent to zh_Hant_TW, which would return true.
-        // We can work around this by searching for both the locale as requested and
-        // the locale in addition to all of its "likely subtags."
-        // This works in practice because, for instance, unum_open("zh_TW") will actually
-        // use "zh_Hant_TW" (confirmed with unum_getLocaleByType(..., ULOC_VALID_LOCALE))
-        // The code below performs, for example, the following mappings:
-        // pa_PK -> pa_Arab_PK
-        // sr_RS -> sr_Cyrl_RS
-        // zh_CN -> zh_Hans_CN
-        // zh_TW -> zh_Hant_TW
-        // TODO(jahorto): Determine if there is any scenario where a language tag + likely subtags is
-        // not exactly functionally equivalent to the language tag on its own -- basically, where
-        // constructor_open(language_tag) behaves differently to constructor_open(language_tag_and_likely_subtags)
-        // for all supported constructors.
-        UErrorCode status = U_ZERO_ERROR;
-        char localeIDWithLikelySubtags[ULOC_FULLNAME_CAPACITY] = { 0 };
-        int likelySubtagLen = uloc_addLikelySubtags(localeID, localeIDWithLikelySubtags, ULOC_FULLNAME_CAPACITY, &status);
-        ICU_ASSERT(status, likelySubtagLen > 0 && likelySubtagLen < ULOC_FULLNAME_CAPACITY);
+        const int count = CountAvailableLocalesFunc();
+        int left = 0;
+        int right = count - 1;
+        int iterations = 0;
+        while (true)
+        {
+            iterations += 1;
+            int i = (left + right) / 2;
+            Assert(i > 0 && i < count);
```

Comment:
seems like i==0 would be valid here and in another assertion below

Reply:
also maybe shouldn't assert this until after we check `left > right`

---
In reply to: [179288298](https://github.com/Microsoft/ChakraCore/pull/4932#discussion_r179288298) [](ancestors = 179288298)